File: backend/src/main/java/com/ll/hotel/domain/member/member/entity/Member.java
package com.ll.hotel.domain.member.member.entity;

import static com.ll.hotel.global.exceptions.ErrorCode.BUSINESS_ACCESS_FORBIDDEN;

import com.ll.hotel.domain.hotel.hotel.entity.Hotel;
import com.ll.hotel.domain.member.member.type.MemberStatus;
import com.ll.hotel.global.jpa.entity.BaseTime;
import com.ll.hotel.global.security.oauth2.entity.OAuth;
import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Table(
    name = "member",
    indexes = {
        @Index(name = "idx_member_email", columnList = "memberEmail")
    }
)
public class Member extends BaseTime {

    @Column(unique = true, nullable = false)
    private String memberEmail;

    @Column(nullable = false)
    private String memberName;

    @Column(nullable = false)
    private String memberPhoneNumber;

    @Column(nullable = true)
    private LocalDate birthDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Role role;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private MemberStatus memberStatus;

    @OneToMany(mappedBy = "member", cascade = CascadeType.ALL)
    @Builder.Default
    private List<OAuth> oauths = new ArrayList<>();

    @OneToOne(mappedBy="member", fetch = FetchType.LAZY)
    private Business business;

    @ManyToMany
    @JoinTable(
        name = "favorite",
        joinColumns = @JoinColumn(name = "member_id"),
        inverseJoinColumns = @JoinColumn(name = "hotel_id")
    )
    @Builder.Default
    private Set<Hotel> favoriteHotels = new HashSet<>();

    public boolean isAdmin() {
        return this.role == Role.ADMIN;
    }

    public boolean isBusiness() {
        return this.role == Role.BUSINESS;
    }

    public boolean isUser() {
        return this.role == Role.USER;
    }

    public String getUserRole() {
        return this.role.name();
    }

    public OAuth getFirstOAuth() {
        return this.oauths.isEmpty() ? null : this.oauths.get(0);
    }

    public void checkBusiness() {
        if (!this.isBusiness()) {
            BUSINESS_ACCESS_FORBIDDEN.throwServiceException();
        }
    }
}

File: backend/src/main/java/com/ll/hotel/domain/member/member/entity/Member.kt
package com.ll.hotel.domain.member.member.entity

import com.ll.hotel.domain.hotel.hotel.entity.Hotel
import com.ll.hotel.domain.member.member.type.MemberStatus
import com.ll.hotel.global.exceptions.ErrorCode.BUSINESS_ACCESS_FORBIDDEN
import com.ll.hotel.global.jpa.entity.BaseTime
import com.ll.hotel.global.security.oauth2.entity.OAuth
import jakarta.persistence.*
import java.time.LocalDate

@Entity
@Table(
    name = "member",
    indexes = [
        Index(name = "idx_member_email", columnList = "memberEmail")
    ]
)
class Member(
    @Column(unique = true, nullable = false)
    var memberEmail: String = "",

    @Column(nullable = false)
    var memberName: String = "",

    @Column(nullable = false)
    var memberPhoneNumber: String = "",

    @Column(nullable = true)
    var birthDate: LocalDate? = null,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var role: Role = Role.USER,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var memberStatus: MemberStatus = MemberStatus.ACTIVE,

    @OneToMany(mappedBy = "member", cascade = [CascadeType.ALL])
    var oauths: MutableList<OAuth> = ArrayList(),

    @OneToOne(mappedBy = "member", fetch = FetchType.LAZY)
    var business: Business? = null,

    @ManyToMany
    @JoinTable(
        name = "favorite",
        joinColumns = [JoinColumn(name = "member_id")],
        inverseJoinColumns = [JoinColumn(name = "hotel_id")]
    )
    var favoriteHotels: MutableSet<Hotel> = HashSet()
) : BaseTime() {

    val isAdmin: Boolean
        get() = this.role == Role.ADMIN

    val isBusiness: Boolean
        get() = this.role == Role.BUSINESS

    val isUser: Boolean
        get() = this.role == Role.USER

    fun getUserRole(): String {
        return this.role.name
    }

    fun getFirstOAuth(): OAuth? {
        return if (this.oauths.isEmpty()) null else this.oauths[0]
    }

    fun checkBusiness() {
        if (!this.isBusiness) {
            BUSINESS_ACCESS_FORBIDDEN.throwServiceException()
        }
    }
} 